pkg/ring/replication_set_test.go

```diff
@@ -33,3 +38,226 @@ func TestReplicationSet_GetAddresses(t *testing.T) {
+				return 1, nil
+			},
+			want: []interface{}{1, 1, 1},
+		},
+		{
+			name: "max unavailable zones = 1, should succeed on instances failing in 1 out of 3 zones (3 instances)",
+			instances: []IngesterDesc{{
+				Zone: "zone1",
+			}, {
+				Zone: "zone2",
+			}, {
+				Zone: "zone3",
+			}},
+			f:                   failingFunctionOnZones("zone1"),
+			maxUnavailableZones: 1,
+			want:                []interface{}{1, 1},
+		},
+		{
+			name: "max unavailable zones = 1, should fail on instances failing in 2 out of 3 zones (3 instances)",
+			instances: []IngesterDesc{{
+				Zone: "zone1",
+			}, {
+				Zone: "zone2",
+			}, {
+				Zone: "zone3",
+			}},
+			f:                   failingFunctionOnZones("zone1", "zone2"),
+			maxUnavailableZones: 1,
+			expectedError:       errZoneFailure,
+		},
+		{
+			name: "max unavailable zones = 1, should succeed on instances failing in 1 out of 3 zones (6 instances)",
+			instances: []IngesterDesc{{
+				Zone: "zone1",
+			}, {
+				Zone: "zone1",
+			}, {
+				Zone: "zone2",
+			}, {
+				Zone: "zone2",
+			}, {
+				Zone: "zone3",
+			}, {
+				Zone: "zone3",
+			}},
+			f:                   failingFunctionOnZones("zone1"),
+			maxUnavailableZones: 1,
+			want:                []interface{}{1, 1, 1, 1},
+		},
+		{
+			name: "max unavailable zones = 2, should fail on instances failing in 3 out of 5 zones (5 instances)",
+			instances: []IngesterDesc{{
+				Zone: "zone1",
+			}, {
+				Zone: "zone2",
+			}, {
+				Zone: "zone3",
+			}, {
+				Zone: "zone4",
+			}, {
+				Zone: "zone5",
+			}},
+			f:                   failingFunctionOnZones("zone1", "zone2", "zone3"),
+			maxUnavailableZones: 2,
+			expectedError:       errZoneFailure,
+		},
+		{
+			name: "max unavailable zones = 2, should succeed on instances failing in 2 out of 5 zones (10 instances)",
+			instances: []IngesterDesc{{
+				Zone: "zone1",
+			}, {
+				Zone: "zone1",
+			}, {
+				Zone: "zone2",
+			}, {
+				Zone: "zone2",
+			}, {
+				Zone: "zone3",
+			}, {
+				Zone: "zone3",
+			}, {
+				Zone: "zone4",
+			}, {
+				Zone: "zone4",
+			}, {
+				Zone: "zone5",
+			}, {
+				Zone: "zone5",
+			}},
+			f:                   failingFunctionOnZones("zone1", "zone5"),
+			maxUnavailableZones: 2,
+			want:                []interface{}{1, 1, 1, 1, 1, 1},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// Ensure the test case has been correctly setup (max errors and max unavailable zones are
+			// mutually exclusive).
+			require.False(t, tt.maxErrors > 0 && tt.maxUnavailableZones > 0)
+
+			r := ReplicationSet{
+				Ingesters:           tt.instances,
+				MaxErrors:           tt.maxErrors,
+				MaxUnavailableZones: tt.maxUnavailableZones,
+			}
+			ctx := context.Background()
+			if tt.cancelContextDelay > 0 {
+				var cancel context.CancelFunc
+				ctx, cancel = context.WithCancel(ctx)
+				go func() {
+					time.AfterFunc(tt.cancelContextDelay, func() {
+						cancel()
+					})
+				}()
```

Comment:
no need for `go func()...` here... `time.AfterFunc` does that already

Reply:
Removed.